staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -596,21 +609,33 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 			growSlice(v, 2048, len(getResp.Kvs))
 		}
 
-		// take items from the response until the bucket is full, filtering as we go
-		for _, kv := range getResp.Kvs {
-			if paging && int64(v.Len()) >= pred.Limit {
-				hasMore = true
-				break
-			}
-			lastKey = kv.Key
+		if paging && pred.Limit < int64(len(getResp.Kvs)) {
+			getResp.Kvs = getResp.Kvs[0:pred.Limit]
+			hasMore = true
+		}
+		// When transforming concurrently, we can't guarantee the order of processing.
```

Comment:
I believe we need to preserve the order of `getResp.Kvs`.

Reply:
Done: preserving the order of Kvs.